File: src/models/basegame.rs
use super::{gameerror::GameError, team::Team};
use serde::{Deserialize, Serialize};

#[derive(Default, Clone, Eq, PartialEq, Debug, Serialize, Deserialize)]
pub struct BaseGame {
    pub name: String,
    pub teams: Vec<Team>,
    pub phases: Vec<super::gamephase::GamePhase>,
    pub current_phase: usize
}

pub trait BaseGameLifeCycle {
    fn name (self: &'_ Self) -> Option<&'_ String>;
    fn start_game(&mut self) -> Result<(), GameError>;
    fn start_phase(&mut self) -> Result<usize, GameError>;
    fn end_phase(&mut self) -> Result<usize, GameError>;
    fn phase_count(&mut self) -> usize;
    fn current_phase(&mut self) -> Result<usize, GameError>;    
    fn end_game(&mut self) -> Result<(), GameError>;
}

impl BaseGame {
	pub fn new() -> BaseGame {
        BaseGame {
            name: String::from("None"),
            phases: Vec::new(),
            teams:  Vec::new(),
            current_phase: 0
        }
    }

    pub fn set_phases(&mut self, phases: Vec<super::gamephase::GamePhase>) {
        for item in phases {
            let _ = &self.phases.push(item);
        }
    }

    pub fn set_teams(&mut self, teams: Vec<super::team::Team>) {
        for item in teams {
            let _ = &self.teams.push(item);
        }
    }

    pub fn phases_count(&mut self) -> usize {
        self.phases.len()
    }

}

impl BaseGameLifeCycle for BaseGame {

    fn name (self: &'_ Self) -> Option<&'_ String> {
        Some(&self.name)
    }

    fn start_game(&mut self) -> Result<(), GameError> {
        
        let phases_count = self.phase_count();
        if phases_count == 0 {
            return Err(GameError::PhasesNotDefined);
        }

        if self.teams.len() < 2 {
            return Err(GameError::TeamsNotDefined);
        }
        self.current_phase = 1;
        Ok(())
    }

    fn phase_count(&mut self) -> usize {
        self.phases.len()
    }

    fn current_phase(&mut self) -> Result<usize, GameError> {
        Ok(self.current_phase)
    }

    fn start_phase(&mut self) -> Result<usize, GameError> {
        Ok(self.current_phase) // TODO
    }

    fn end_phase(&mut self) -> Result<usize, GameError> {
        let phase_len = self.phases.len();

        if self.current_phase==0 {
            return Err(GameError::GameNotStarted);
        }

        log::error!("phase_len {:?}", phase_len);
        if self.current_phase > phase_len {
            let _msg = format!("The game is already finished");
            Err(GameError::GameAlreadyFinished)
        }
        else if self.current_phase == phase_len {
            self.current_phase += 1;
            Ok(phase_len)
        }
        else {
            self.current_phase += 1;
            Ok(self.current_phase)
        }
    }

    fn end_game(&mut self) -> Result<(), GameError> {
        Ok(())
    }


   
}

File: java/src/main/java/com/foul/gameserver/domain/base/BaseGame.java
package com.foul.gameserver.domain.base;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.List;
import com.foul.gameserver.domain.base.result.GameResult;
import com.foul.gameserver.domain.base.result.PhaseResult;
import com.foul.gameserver.exception.GameError;
import com.foul.gameserver.exception.GameErrorException;
import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.Setter;
import lombok.ToString;

@Getter
@EqualsAndHashCode
@ToString
public class BaseGame implements BaseGameLifeCycle, BaseGameInfo {

    Logger logger = LoggerFactory.getLogger(BaseBoard.class);
    
    @Setter
    protected String name;

    @Setter
    protected List<Team> teams;

    @Setter
    protected List<GamePhase> phases;

    @Setter
    protected GamePhase currentPhase;

    @Setter
    protected int currentPhaseNumber;

    @Setter
    protected Player currentPlayer;

    @Setter
    protected BaseBoard gameBoard;

    public BaseGame() {
        this.currentPhaseNumber = 0;
    }

    protected int getMinTeamsSize() {
        return 1;
    }

    @Override
    public GameResult startGame() throws GameErrorException {
        logger.debug("BaseGame::startGame::start");

        var phaseCount = this.phaseCount();
        if (phaseCount == 0) {
            throw new GameErrorException(GameError.PhasesNotDefined);
        }

        if (this.teams.size() < this.getMinTeamsSize()) {
            throw new GameErrorException(GameError.TeamsNotDefined);
        }
        this.currentPhaseNumber = 1;
        this.currentPhase = this.getNextPhase(null);

        logger.debug("BaseGame::startGame::end");
        // TODO Phase real
        return new GameResult();
    }

    @Override
    public PhaseResult startPhase() throws GameErrorException {
        logger.debug("BaseGame::startPhase::start");
        return new PhaseResult(); // TODO
    }

    @Override
    public PhaseResult endPhase() throws GameErrorException {
        logger.debug("BaseGame::endPhase::start");
        if (this.currentPhase() == null) {
            throw new GameErrorException(GameError.GameNotStarted);
        }

        var phaseLen = this.phaseCount();
        if (this.currentPhaseNumber() > phaseLen) {
            throw new GameErrorException(GameError.GameAlreadyFinished);
        } else if (this.currentPhaseNumber() == phaseLen) {
            this.currentPhaseNumber += 1;
            return new PhaseResult(); // TODO
        } else {
            this.currentPhaseNumber += 1;
            return new PhaseResult(); // TODO
        }
    }

    public BaseBoard getBoard() throws GameErrorException {
        logger.debug("BaseGame::getBoard");
        if (this.gameBoard == null) {
            throw new GameErrorException("No board defined");
        }
        return this.gameBoard;
    }

    @Override
    public GameResult endGame() throws GameErrorException {
        logger.debug("BaseGame::endGame");
        return null;
    }

    @Override
    public int phaseCount() {
        logger.debug("BaseGame::phaseCount " + this.phases.size());
        return this.phases.size();
    }

    @Override
    public GamePhase currentPhase() {
        logger.debug("BaseGame::currentPhase " + this.currentPhase);
        return this.currentPhase;
    }

    @Override
    public int currentPhaseNumber() {
        logger.debug("BaseGame::currentPhaseNumber " + this.currentPhaseNumber);
        return this.currentPhaseNumber;
    }

    @Override
    public Player currentPlayer() {
        logger.debug("BaseGame::currentPlayer " + this.currentPlayer);
        return this.currentPlayer;
    }

    private GamePhase getNextPhase(GamePhase phase) {
        logger.debug("BaseGame::getNextPhase");
        if (phase == null) {
            return this.phases.get(0);
        }

        return this.phases.stream().filter(item -> phase.getUid().equals(item.getUid())).findFirst()
                .orElse(null);
    }

}
